### ISSUE TITLE
Ground Texture Low Quality

### ISSUE THREAD

**Issue by dol-leodagan** [Merec/DAoC-MapCreator#2]
Hello Merec,

I'm rendering some map at really high resolution (4096~8192 maybe above when I can get more Memory !)

Terrain texture seems really low quality when zoomed in, where fixture (and shaded trees) are nicely drawn even in high level zoom...

Are there any chance to get interpolated textures or other mean of getting better ground texture ?

This could be needed in tile-based Map browser... (Like Open Street Map)

**Comment by dol-leodagan** [comment]
Adding this line at MapBackground.cs, line 122 allow to add interpolation algorithm :
```
                        mapTile.Interpolate = PixelInterpolateMethod.Bilinear;
```
This could need further testing or even some drop-down menu to allow user selection...

**Comment by dol-leodagan** [comment]
After Testing, changing the Interpolate algorithm has little to no effect...

Other Options are changing FilterType (and there is a lot of Filter and test to do)

Changing Sharpen ratio, or using Unsharpmask method...

And also try other method like Enhance to reduce noise... This can be done Tile by Tile or Zone wide, after or before sharpening...

I'm also trying other lightmap composition operator, some are giving nice shadows, but overall lighting is low ;)

I'm not quite good in image processing !

**Comment by Merec** [comment]
I guess the problem is not the image scaling, its the source itself. Checkout the "tex*" or "lod*" files in the zone you are creating. There is the problem.
 
![tex00-00](https://cloud.githubusercontent.com/assets/3285379/10566475/1a5d6a24-75e8-11e5-9d11-de462cee7476.jpg)

It is possible to generate new terrain tiles using the "patch"-Files within the zone directory, but that need some more investigation.

**Comment by dol-leodagan** [comment]
Yes that is the "noise" I'm seeing on high resolution rendering...

**Comment by Merec** [comment]
After updating to latest Magick.Net (see #3) I will have a look at better tile generation not using the prerendered images.

**Comment by dol-leodagan** [comment]
The game engine seem to use some texture mapping also, there are "common" textures in the Zones directory which are referenced by some CSV files within each zone MPK... (like mapping grass texture or rock texture)

And each Zone seem to have a 1024x1024 terrain texture for the whole area (maybe used by scaling and blending with each sub zone)

And there are also some shadow map which could help with the lighting made in MapCreator which is only using the heightmap normals for now...

**Comment by Merec** [comment]
I have updated the project, it now uses the latest Magick.NET version and NuGet. Lets see, how to enhance the output.

**Comment by dol-leodagan** [comment]
For now the only difference I see is that shadow are more diffuse (maybe something new around the lighting filters or the sharpen filter that balance brightness)

I can't make sure I used the exact same setting, but I don't change them a lot around lights...

Texture is still noisy... but I think that's still because we use only part of the real texturing process.

I'm working on a client "Terrain" renderer actually, I don't use texture, but I wonder if it could be better to stretch it according the real terrain height before using it as flat topview...

**Comment by Merec** [comment]
Maybe this little quote helps you:

> .pcx with each r=0 being the first layer, r=1 the 2nd, ... (water00, water01 in sector.dat)
> or for textures?
> textures is those patchXXXX-YY files
> XXXX = (x,y) subtile, 8x8 per zone
> yy = id
> for each subtile you basically go through (r,g,b)
> r= first texture
> g = 2nd texture ...
> r of -01 = 4th layer
> and so on
> so you can stack it indefinitely. it's always alpha-blended with the previous layer
> i.e. r=128 means i'll apply the current layer with 50% alpha ontop of what is already there
> base of everything is the original SI lod map
> (i.e. if i have a layer of r=0 only, you'lll see LOD)
> texture mapping is in the textures.csv in the patch dir ..
> there
> there's a "tile" value there which is 1/n or something
> 256 means "repeat this texture 256 times in this subtile"

**Comment by dol-leodagan** [comment]
Yeah that's a lot of work to implement, I understand now how the "Level of Details" from Client terrain work ;)

Anyway I think you have some error in your height map...

in MapHeightmap.cs around lines 85 to 88

                                ushort terrainPixelValue = (ushort)(terrainPixels[x, y].GetChannel(0) / 256);
                                ushort offsetPixelValue = (ushort)(offsetPixels[x, y].GetChannel(0) / 256);
                                ushort heightmapPixelValue = (ushort)(terrainPixelValue * m_terrainfactor + offsetPixelValue * m_offsetfactor);

Should be 

                                float terrainPixelValue = terrainPixels[x, y].GetChannel(0) / 256f;
                                float offsetPixelValue = offsetPixels[x, y].GetChannel(0) / 256f;
                                ushort heightmapPixelValue = (ushort)(terrainPixelValue * m_terrainfactor + offsetPixelValue * m_offsetfactor);

Offset and Terrain Map are 8bit grayscale, which mean there is only 255 values by pixel, if you scale it to 256 you'll have a result always under 1.0, then converted to any integer value it will be rounded to 0 or 1... you miss a lot of precision

Then again you convert a scaled "client" Piexel value (with 8 scale and 32 offset most of the time)  which offer about 40 different value, to a grayscale "height" map holding 16bits per components...

**Comment by Merec** [comment]
Thanks for the hint! I guess it has something to do with the magick versions, the older version does not use ushort values which was consfusing sometimes and at one point I got a little upset about math :P

Feel free to create fork and submit a merge request, thats much easier to test it for me.

**Comment by dol-leodagan** [comment]
Well I'm not modifying your work at all...

I'm using some of your methods to build a 3D view of terrain, I'm starting a "collision" map project !

**Comment by Merec** [comment]
I have testet what you mentioned. I am unable to see any differences in the results. I also set the divider from 256 to 255.

**Comment by Merec** [comment]
So, is there any way to increase the quality of ground textures?

**Comment by dol-leodagan** [comment]
The higher quality textures need to be interpolated from what I understand on how the client draw the terrain...

The fix I gave should mostly impact lighting and shadows.
